tc/core/tensor.cc

```diff
@@ -0,0 +1,123 @@
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include <version.h>
+
+#include <cstdint>
+#include <fstream>
+#include <numeric>
+#include <tuple>
+
+#include "tc/core/tensor.h"
+#include "tc/core/utils/math.h"
+
+namespace tc {
+
+namespace {
+uint64_t GetDLTensorAlignment(const DLTensor* t) {
+  return (reinterpret_cast<std::uintptr_t>(t->data) + t->byte_offset) % 256;
+}
+uint64_t GetDLTensorAlignment(const DLConstTensor* t) {
+  return (reinterpret_cast<std::uintptr_t>(t->data) + t->byte_offset) % 256;
+}
+} // namespace
+
+TensorInfo::TensorInfo(
+    DLDataType t,
+    uint64_t align,
+    const std::vector<int64_t>& sz,
+    const std::vector<int64_t>& st)
+    : dtype(t), alignment(align), shape(sz), strides(st) {}
+
+TensorInfo::TensorInfo(const DLTensor* t)
+    : dtype(t->dtype), alignment{GetDLTensorAlignment(t)} {
+  shape.reserve(t->ndim);
+  std::copy(t->shape, t->shape + t->ndim, std::back_inserter(shape));
+  if (not t->strides) {
+    return;
+  }
+  strides.reserve(t->ndim);
+  std::copy(t->strides, t->strides + t->ndim, std::back_inserter(strides));
+}
+
+TensorInfo::TensorInfo(const DLConstTensor* t)
+    : dtype(t->dtype), alignment{GetDLTensorAlignment(t)} {
+  shape.reserve(t->ndim);
+  std::copy(t->shape, t->shape + t->ndim, std::back_inserter(shape));
+  if (not t->strides) {
+    return;
+  }
+  strides.reserve(t->ndim);
+  std::copy(t->strides, t->strides + t->ndim, std::back_inserter(strides));
+}
+
+TensorInfo::TensorInfo(const TensorInfoProto& buf)
+    : dtype{static_cast<uint8_t>(buf.dtype().code()),
+            static_cast<uint8_t>(buf.dtype().bits()),
+            static_cast<uint16_t>(buf.dtype().lanes())},
+      alignment{buf.alignment()},
+      shape{buf.shape().begin(), buf.shape().end()},
+      strides{buf.strides().begin(), buf.strides().end()} {}
+
+TensorInfoProto TensorInfo::toProtobuf() const {
+  TensorInfoProto buf;
+  buf.mutable_shape()->Reserve(shape.size());
+  std::copy(
+      shape.begin(),
+      shape.end(),
+      google::protobuf::RepeatedFieldBackInserter(buf.mutable_shape()));
+  buf.mutable_strides()->Reserve(strides.size());
+  std::copy(
+      strides.begin(),
+      strides.end(),
+      google::protobuf::RepeatedFieldBackInserter(buf.mutable_strides()));
+  buf.set_alignment(alignment);
+  buf.mutable_dtype()->set_code(dtype.code);
+  buf.mutable_dtype()->set_bits(dtype.bits);
+  buf.mutable_dtype()->set_lanes(dtype.lanes);
+  return buf;
+}
+
+bool operator==(const DLDataType& a, const DLDataType& b) {
+  return a.code == b.code and a.bits == b.bits and a.lanes == b.lanes;
+}
+
+bool operator<(const DLDataType& a, const DLDataType& b) {
```

Comment:
Not sure default less-comparison operators make sense for types.  Why do we need them, to use types in ordered containers?  I would prefer function objects `DLDataTypeSimpleCompare`.